dplyr 1.0.4: if_any(), if_all(), faster across() (#497)
tidyverse/tidyverse.org: content/blog/dplyr-1-0-4-if-any/index.Rmd

date: 2021-02-01
author: Romain Francois
description: >
    Minor release with major performance improvements for `across()` and 
    two new functions `if_any()` and `if_all()`.

photo:
  url: https://unsplash.com/photos/nQz49efZEFs
  author: Mattias Olsson

categories: [package] 
tags: [dplyr]
---

We're happy to announce the release of [dplyr](https://dplyr.tidyverse.org) 1.0.4, 
discussed in this blog post. The highlights of this release are: two new functions `if_all()` and `if_any()`, and improved performance improvements of `across()`. 

You can install it from CRAN with:

```{r, eval = FALSE}
install.packages("dplyr")
```

You can see a full list of changes in the [release notes](https://github.com/tidyverse/dplyr/releases/tag/v1.0.4). 

```{r setup, echo=FALSE}
library(dplyr, warn.conflicts = FALSE)
```

## if_any() and if_all()

The new `across()` function introduced as part of 
[dplyr 1.0.0](https://www.tidyverse.org/blog/2020/04/dplyr-1-0-0-colwise/) is proving to be 
a successful addition to dplyr. In case you missed it, `across()` lets you 
conveniently express a set of actions to be performed across a tidy selection 
of columns. 

> `across()` is very useful within `summarise()` and `mutate()`, but it can be confusing when used in `filter()` because it is not clear how the results can be combined into one logical vector. 

Reviewer: ```suggestion
`across()` is very useful within `summarise()` and `mutate()`, but it's hard to use it with `filter()` because it is not clear how the results would be combined into one logical vector. 
```

So to fill the gap, we're introducing two new functions `if_all()` and `if_any()`. Let's 
directly dive in to an example: 

```{r}
library(palmerpenguins)

big <- function(x) {
  x > mean(x, na.rm = TRUE)
}

# keep rows if all the selected columns are "big"
penguins %>% 
  filter(if_all(contains("bill"), big))

# keep rows where at least one of the columns is "big"
penguins %>% 
  filter(if_any(contains("bill"), big))
```

Both functions operate similarly to `across()` but go the extra mile of aggregating the 
results to indicate if *all* the results are true when using `if_all()`, or if *at least 
one* is true when using `if_any()`. 

Although `if_all()` and `if_any()` were designed with `filter()` in mind, we 
[then discovered](https://github.com/tidyverse/dplyr/issues/5709) that they can also be

> useful within `mutate()` and/or `summarise()`. 

Reviewer: ```suggestion
useful within `mutate()` and/or `summarise()`: 
```

```{r}
penguins %>% 
  filter(!is.na(bill_length_mm)) %>% 
  mutate(
    category = case_when(
      if_all(contains("bill"), big) ~ "both big", 
      if_any(contains("bill"), big) ~ "one big", 
      TRUE                          ~ "small"
    )) %>% 
  count(category)
```

## Faster across()

One of the main motivations for across() was eliminating the need for 

> every verb to have a `_at`, `_if`, and `_all` variant. Unfortunately, however, this came with a performance cost. In this release, we have redesigned `across()` to eliminate that problem when possible. 

Reviewer: ```suggestion
every verb to have a `_at`, `_if`, and `_all` variant. Unfortunately, however, this came with a performance cost. In this release, we have redesigned `across()` to eliminate that performance penalty in many cases. In the following example, you can now see that the old and new approaches take the same amount of time.
```

```{r}
library(vroom)

mun2014 <- vroom("https://data.regardscitoyens.org/elections/2014_municipales/MN14_Bvot_T1_01-49.txt", 

>                  col_select = -c('X4','X9','X10','X11'),

Reviewer: Add `col_types = list()` to suppress the type output

Author: Thanks

    two new functions if_any() and if_all().

photo:
  url: https://unsplash.com/photos/nQz49efZEFs
  author: Mattias Olsson

categories: [package] 
tags: [dplyr]
---

We're happy to announce the release of [dplyr](https://dplyr.tidyverse.org) 1.0.4. dplyr is the data
manipulation package of the [tidyverse](https://www.tidyverse.org). 

This release dramatically improves the performance of `across()` and introduces 
the new functions `if_all()` and `if_any()` as companions to `filter()`. 
You can install it from CRAN with:

```{r, eval = FALSE}
install.packages("dplyr")
```

This blog post will discuss the two most important changes of the 1.0.4 release: 
the new `if_any()` and `if_all()` functions, and performance improvements for `across()`.

You can see a full list of changes in the [release notes]({ github_release })

```{r setup}
library(dplyr, warn.conflicts = FALSE)
```

## if_any() and if_all()

The new `across()` function introduced as part of dplyr 1.0.0 is starting to be 
a successful addition to dplyr. In case you missed it, `across()` lets you 
conveniently express a set of actions to be performed across a tidy selection 
of columns. 

> However useful `across()` may be within `summarise()` and `mutate()`, we discovered

Author: I've tried to avoid explaining in terms of across() because we want to make across() warn when used in filter() later

> However useful `across()` may be within `summarise()` and `mutate()`, we discovered

Reviewer: Ah good point. Maybe something more like:  `across()` is very useful within `summarise()` and `mutate()`, but while it's technically possible to use within `filter()`, it's not at all obvious what it might do.

useful when used within `mutate()` and/or `summarise()`. 

```{r}
penguins %>% 
  filter(!is.na(bill_length_mm)) %>% 
  mutate(
    category = case_when(
      if_all(contains("bill"), big) ~ "both big", 
      if_any(contains("bill"), big) ~ "one big", 
      TRUE                          ~ "small"
    )) %>% 
  count(category)
```

## Faster across()

One net benefit of `across()` is that it supersedes a collection of functions, 
namely the column wise set of functions (`summarise_at()` ...). However, up until the 
previous release, this came with a cost in terms of performance. 

In the 1.0.4 release, we have redesigned `across()` to perform better by doing less
when possible. With these improvements, using `across()` within `summarise()` or `mutate()`
has the same performance as the relevant column wise version. 

```{r}
library(vroom)

mun2014 <- vroom("https://data.regardscitoyens.org/elections/2014_municipales/MN14_Bvot_T1_01-49.txt", 
                 col_select = -c('X4','X9','X10','X11'),
                 col_names = FALSE, 
                 locale = locale(encoding = "WINDOWS-1252"), altrep = FALSE) 
bench::workout({

>   a <- mun2014 %>% group_by_if( is.character )

Reviewer: Spacing looks a little weird here.
